samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaConsumerProxy.java

```diff
@@ -0,0 +1,447 @@
+      SystemStreamPartition ssp = e.getValue();
+      if (systemStreamPartitions.contains(ssp)) {
+        topicPartitionsToPoll.add(tp);  // consume
+      } else {
+        topicPartitionsToPause.add(tp); // ignore
+      }
+    }
+
+    ConsumerRecords<K, V> records;
+    // make a call on the client
+    try {
+      // Currently, when doing checkpoint we are making a safeOffset request through this client, thus we need to synchronize
+      // them. In the future we may use this client for the actually checkpointing.
+      synchronized (kafkaConsumer) {
+        // Since we are not polling from ALL the subscribed topics, so we need to "change" the subscription temporarily
+        kafkaConsumer.pause(topicPartitionsToPause);
+        kafkaConsumer.resume(topicPartitionsToPoll);
+        records = kafkaConsumer.poll(timeout);
+        // resume original set of subscription - may be required for checkpointing
+        kafkaConsumer.resume(topicPartitionsToPause);
+      }
+    } catch (InvalidOffsetException e) {
+      // If the consumer has thrown this exception it means that auto reset is not set for this consumer.
+      // So we just rethrow.
+      LOG.error("Caught InvalidOffsetException in pollConsumer", e);
+      throw e;
+    } catch (KafkaException e) {
+      // we may get InvalidOffsetException | AuthorizationException | KafkaException exceptions,
+      // but we still just rethrow, and log it up the stack.
+      LOG.error("Caught a Kafka exception in pollConsumer", e);
+      throw e;
+    }
+
+    return processResults(records);
+  }
+
+  private Map<SystemStreamPartition, List<IncomingMessageEnvelope>> processResults(ConsumerRecords<K, V> records) {
+    if (records == null) {
+      throw new SamzaException("processResults is called with null object for records");
+    }
+
+    int capacity = (int) (records.count() / 0.75 + 1); // to avoid rehash, allocate more then 75% of expected capacity.
+    Map<SystemStreamPartition, List<IncomingMessageEnvelope>> results = new HashMap<>(capacity);
+    // Parse the returned records and convert them into the IncomingMessageEnvelope.
+    // Note. They have been already de-serialized by the consumer.
+    for (ConsumerRecord<K, V> record : records) {
+      int partition = record.partition();
+      String topic = record.topic();
+      TopicPartition tp = new TopicPartition(topic, partition);
+
+      updateMetrics(record, tp);
+
+      SystemStreamPartition ssp = topicPartitions2SSP.get(tp);
+      List<IncomingMessageEnvelope> listMsgs = results.get(ssp);
+      if (listMsgs == null) {
+        listMsgs = new ArrayList<>();
+        results.put(ssp, listMsgs);
+      }
+
+      final K key = record.key();
+      final Object value = record.value();
+      final IncomingMessageEnvelope imEnvelope =
+          new IncomingMessageEnvelope(ssp, String.valueOf(record.offset()), key, value, getRecordSize(record));
+      listMsgs.add(imEnvelope);
+    }
+    if (LOG.isDebugEnabled()) {
```

Comment:
Why do we have  the log level guard here. In slf4j/log4j combination, event can be logged unconditionally and message formatting is applied only based upon the configured  log level of the application. To me this check introduces unnecessary complexity, what is the value add we get from this?

Reply:
In this case we need it (because there is a loop), but in other cases I've removed it.